### PR TITLE
[Merged by Bors] - feat(Set,Finset): add lemmas, golf

### PR DESCRIPTION
**Pull request by urkud**
## New lemmas

- `Set.image_nontrivial`
- `Finset.map_ssubset_map`, `Finset.map_nontrivial`

## New `gcongr` attrs

- `Finset.map_subset_map`, `Finset.map_ssubset_map` (`.mpr` implications);
- `Finset.product_subset_product_left`, `Finset.product_subset_product_right`

## Misc

- `Finset.Nonempty.map` is now `protected`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Image.lean [hunk 1]
```diff
@@ -1238,6 +1238,10 @@ theorem nontrivial_of_image (f : α → β) (s : Set α) (hs : (f '' s).Nontrivi
   ⟨x, hx, y, hy, mt (congr_arg f) hxy⟩
 #align set.nontrivial_of_image Set.nontrivial_of_image
 
+@[simp]
+theorem nontrivial_image {f : α → β} (hf : f.Injective) : (f '' s).Nontrivial ↔ s.Nontrivial :=
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Image.lean:1242]:
```suggestion
theorem image_nontrivial {f : α → β} (hf : f.Injective) : (f '' s).Nontrivial ↔ s.Nontrivial :=
```

File: Mathlib/Data/Finset/Prod.lean [hunk 2]
```diff
@@ -93,10 +93,12 @@ theorem product_subset_product (hs : s ⊆ s') (ht : t ⊆ t') : s ×ˢ t ⊆ s'
   mem_product.2 ⟨hs (mem_product.1 h).1, ht (mem_product.1 h).2⟩
 #align finset.product_subset_product Finset.product_subset_product
 
+@[gcongr]
 theorem product_subset_product_left (hs : s ⊆ s') : s ×ˢ t ⊆ s' ×ˢ t :=
   product_subset_product hs (Subset.refl _)
 #align finset.product_subset_product_left Finset.product_subset_product_left
 
+@[gcongr]
```
Comment (YaelDillies) [inline Mathlib/Data/Finset/Prod.lean:101]: Should that not go on `product_subset_product` instead? Not sure what's preferred but I had gathered it was tagging the most general lemma.

Comment (urkud) [reply Mathlib/Data/Finset/Prod.lean:101]: With `gcongr`, you need to tag both the general lemma and `left`/`right` lemmas. You can think of it as a workaround for a ~bug~missing feature in `gcongr`.

File: Mathlib/Data/Finset/Image.lean [hunk 3]
```diff
@@ -278,19 +282,21 @@ theorem map_cons (f : α ↪ β) (a : α) (s : Finset α) (ha : a ∉ s) :
...
+  mod_cast Set.image_nonempty (f := f) (s := s)
 #align finset.map_nonempty Finset.map_nonempty
 
-alias ⟨_, Nonempty.map⟩ := map_nonempty
+protected alias ⟨_, Nonempty.map⟩ := map_nonempty
 #align finset.nonempty.map Finset.Nonempty.map
 
+@[simp]
+theorem map_nontrivial : (s.map f).Nontrivial ↔ s.Nontrivial :=
+  mod_cast Set.nontrivial_image f.injective (s := s)
```
Comment (urkud) [inline Mathlib/Data/Finset/Image.lean:298]:
```suggestion
  mod_cast Set.image_nontrivial f.injective (s := s)
```

**Review by YaelDillies — APPROVED** [review]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**